File: src/media.rs
extern crate image;

use std::path;
use std::io::Write;

use image::GenericImage;
use image::FilterType;
use image::ImageFormat;


#[derive(RustcEncodable)]
pub struct Media {
    name: String,
    dimensions: (u32, u32),
}


pub fn path_to_media(path: path::PathBuf) -> Option<Media> {
    let img = match image::open(&path) {
        Ok(i) => i,
        Err(_) => return None,
    };

    let name = match path.as_path()
        .file_name()
        .and_then(|e| e.to_str())
        .map(|e| e.to_string()) {
            Some(n) => n,
            None => return None,
        };

    Some(Media{
        name: name,
        dimensions: img.dimensions(),
    })
}


#[test]
fn test_path_to_media_1() {
    let media = path_to_media(path::Path::new("src/test/1.jpeg").to_path_buf()).unwrap();

    assert!(media.dimensions == (225, 225));
}

#[test]
fn test_path_to_media_2() {
    let media = path_to_media(path::Path::new("src/test/2.jpeg").to_path_buf()).unwrap();

    assert!(media.dimensions == (107, 182));
}

#[test]
fn test_path_to_media_3() {
    let media = path_to_media(path::Path::new("src/test/2016/3.jpeg").to_path_buf()).unwrap();

    assert!(media.dimensions == (225, 225));
}


pub fn resize_media<W: Write>(path: path::PathBuf, width: u32, writer: &mut W) {
    let image = match image::open(&path) {
        Ok(i) => i,
        Err(_) => return,
    };

    let (owidth, oheight) = image.dimensions();

    let height = width * oheight / owidth;

    let resized = image.resize(width, height, FilterType::Triangle);

    resized.save(writer, ImageFormat::JPEG);
}

File: backend/media.py
import os
from enum import Enum
from PIL import Image

class MediaType(Enum):
    unknown = 0
    image = 1


class Media(object):
    def __init__(self, name, realpath, album):
        self.name = name
        self.realpath = realpath
        self.album = album

        try:
            image = Image.open(realpath)
            self.width = image.width
            self.height = image.height
            self.type = MediaType.image
        except:
            self.type = MediaType.unknown


    def is_unknown(self):
        return self.type == MediaType.unknown

    def to_dict(self):
        return {
                'name': self.name,
                'width': self.width,
                'height': self.height,
                'path': os.path.join(self.album.path, self.name),
                }
